### ISSUE TITLE
tensorflow.python.framework.errors_impl.AlreadyExistsError: TensorFlow device (GPU:0) is being mapped to multiple CUDA devices (0 now, and 1 previously), which is not supported. This may be the result of providing different GPU configurations (ConfigProto.gpu_options, for example different visible_device_list) when creating multiple Sessions in the same process. This is not  currently supported, see https://github.com/tensorflow/tensorflow/issues/19083

### ISSUE THREAD

**Issue by Nafees-060** [horovod/horovod#3653]
**Environment:**
1. Framework: (TensorFlow)
2. Framework version: 2.2.0
3. Horovod version:  v0.21.3
4. MPI version:  (Open MPI) 2.1.1
5. CUDA version: 10.1, V10.1.243
6. NCCL version:  2.11.4
7. Python version: 3.6.9
11. CMake version:  3.10.2

**Bug report:**
I am trying to utilize the multi-GPUs using Horovod for distributed training.  Initially, I utilized a single GPU and two GPUs to test a simple convolution neural network. Everything functions properly. Then, I used CNN and LSTM in combination. It works perfectly on a single GPU; however, a problem arises when running on two GPUs. The complete Trackback is as follows:
```
[1,1]<stderr>:Traceback (most recent call last):
[1,1]<stderr>:  File "horovod-PAMAP2.py", line 203, in <module>
[1,1]<stderr>:    PAMAP2()
[1,1]<stderr>:  File "horovod-PAMAP2.py", line 144, in PAMAP2
[1,1]<stderr>:    model = deep_Model(input_shape, num_classes)
[1,1]<stderr>:  File "/research/dept8/gds/anafees/HAR-MGDP/horo_Model.py", line 30, in deep_Model
[1,1]<stderr>:    lstm_fm = LSTM_2016(con_l4)
[1,1]<stderr>:  File "/research/dept8/gds/anafees/HAR-MGDP/horo_Model.py", line 15, in LSTM_2016
[1,1]<stderr>:    recurrent_dropout=0.5, return_sequences=True)(
[1,1]<stderr>:  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/layers/recurrent_v2.py", line 1096, in __init__
[1,1]<stderr>:    if context.num_gpus() > 0:
[1,1]<stderr>:  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/context.py", line 2046, in num_gpus
[1,1]<stderr>:    return context().num_gpus()
[1,1]<stderr>:  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/context.py", line 1047, in num_gpus
[1,1]<stderr>:    self.ensure_initialized()
[1,1]<stderr>:  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/context.py", line 515, in ensure_initialized
[1,1]<stderr>:    context_handle = pywrap_tfe.TFE_NewContext(opts)
[1,1]<stderr>:tensorflow.python.framework.errors_impl.AlreadyExistsError: TensorFlow device (GPU:0) is being mapped to multiple CUDA devices (0 now, and 1 previously), which is not supported. This may be the result of providing different GPU configurations (ConfigProto.gpu_options, for example different visible_device_list) when creating multiple Sessions in the same process. This is not  currently supported, see https://github.com/tensorflow/tensorflow/issues/19083
-------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code.. Per user-direction, the job has been aborted.
-------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[49535,1],1]
```

**Comment by maxhgerlach** [comment]
Did you pin the GPU to be used to the process local rank (one GPU per process)?  https://horovod.readthedocs.io/en/stable/summary_include.html#usage

**Comment by Nafees-060** [comment]
@maxhgerlach  thank you so much for response. Yes  I pinned the GPU... Here is the code:
```
# Horovod: pin GPU to be used to process local rank (one GPU per process)

hvd.init()
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
config.gpu_options.visible_device_list = str(hvd.local_rank())
K.set_session(tf.compat.v1.Session(config=config))
```

**Comment by maxhgerlach** [comment]
Have you tried to do that the TF2 way using `tf.config.set_visible_devices()`? https://www.tensorflow.org/api_docs/python/tf/config/set_visible_devices

**Comment by Nafees-060** [comment]
@maxhgerlach can you please elaborate more about TF2 way using?  I mean what should I put as an argument in `tf.config.set_visible_devices()`, and is there anything else needs to utilize as an in TF2.

**Comment by maxhgerlach** [comment]
Something like
```
physical_gpus = tf.config.list_physical_devices('GPU')
tf.config.set_visible_devices([physical_gpus[hvd.local_rank()]], "GPU")
```

**Comment by Nafees-060** [comment]
@maxhgerlach  it worked. Thanks alot. But how it works, can you explain. Thank you

**Comment by maxhgerlach** [comment]
Great! Not sure what 
```
K.set_session(tf.compat.v1.Session(config=config))
```
did in your original code, but apparently not the right / sufficient thing. :)

**Comment by Nafees-060** [comment]
@maxhgerlach Thank you so much for the help. One more question, I am implementing some **time series classification using horovod on multi-GPUs.** I initially used a simple 4-layers convolutional neural network and tested training time on single and double GPUs with 4  CPUs. I have seen a big overhead on two GPUs compared to a single. Then I tried increasing parameters using LSTM and got the same result. Then I tried to increase the data, still got the same issue. I am confused about whether multi-GPUs are not efficient for training or if I am doing something wrong in implementation or utilization.
